### PR TITLE
Release automation - Run smoke tests when zip file is created

### LINKED ISSUE #5528: Release Automation - Run E2E tests after building zip file
### Description

At the moment, smoke tests are run manually against the zip file created as part of release automation process. To improve dev efficiency and reduce incidents, we can run E2E tests against the zip file created, which includes most of the tests ran as part of smoke tests.

### Acceptance criteria

E2E tests should run immediately after the zip file is created as part of release automation.

### PR DESCRIPTION
**Pull request by achyuthajoy**
Fixes #5528 

### Changes proposed in this Pull Request

The idea behind this PR is to automatically run smoke tests when a release zip is built. Since we don't have dedicated smoke tests suite at the moment, we'll be running the E2E - Pull Request test suite against the release zip file until dedicated smoke tests are available.

This PR will:
1. Run smoke tests with [build zip file workflow](https://github.com/Automattic/woocommerce-payments/actions/workflows/build-zip-file.yml). Allows skipping smoke tests if required.
2. Builds the zip file & run smoke tests with [Release - Create PR to trunk](https://github.com/Automattic/woocommerce-payments/blob/develop/.github/workflows/release-pr.yml). Allows skipping smoke tests if required.

#### Testing Build zip file  workflow

* Trigger a new manual action for [Build zip file](https://github.com/Automattic/woocommerce-payments/actions/workflows/build-zip-file.yml).
* Confirm that a checkbox `Skip running smoke tests` is displayed when the branch is selected.
* Trigger a new test against this branch without checking `Skip running smoke tests` and confirm that E2E - Pull request tests are run after build zip job is complete.
* Trigger a new test against this branch with `Skip running smoke tests` checked and confirm that E2E - Pull request tests are skipped.

#### Testing Release - Create PR to trunk  workflow

* Trigger a new manual action for [Release - Create PR to trunk](https://github.com/Automattic/woocommerce-payments/blob/develop/.github/workflows/release-pr.yml).
* Confirm that a checkbox `Skip running smoke tests` is displayed when the branch is selected.
* Trigger a new test against this branch with a specific release version & without checking `Skip running smoke tests`. Confirm that E2E - Pull request tests are run after build zip job is complete.
* Trigger a new test against this branch with a specific release version & `Skip running smoke tests` checked. Confirm that E2E - Pull request tests are skipped.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**
- [ ] Update release train model to include the changes introduced on this PR - @achyuthajoy
- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [inline .github/workflows/build-zip-file.yml]: In which use case would we want to skip building the zip file within the `build-zip-file` workflow?
It feels like we don't need such input

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [inline .github/workflows/build-zip-file.yml]: We can just checkout the relevant branch in the caller workflows before calling this one (and remove the `ref` param l.37/38). If we can avoid piling up inputs (and having to propagate some to child workflows) , that would be best in my opinion

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [inline .github/workflows/build-zip-file.yml]: Is `if: ${{ inputs.skip-smoke-tests != true }}` not enough?

Since the job depends on `build-zip`, which is skipped when `inputs.skip-build-zip != true`

if job B depends on job A, I would expect that job B is skipped automatically if job A is also skipped. Is that not the case?

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [inline .github/workflows/build-zip-file.yml]: Is this really needed?

I think the `npm run build` command already does that, so I'm not sure:
https://github.com/Automattic/woocommerce-payments/blob/d276e3a4efed3d380377cee9d33356ed541b265e/package.json#L23

File: .github/workflows/create-pre-release.yml
Comment (jbordonado) [inline .github/workflows/create-pre-release.yml]: I thought we agreed that it was not necessary to run the smoke tests for a pre-release package, given that we're running all our E2E tests on `develop` branch anyway, and the only change we're doing on the fly before building the package is adjust the version number in `package.json`

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [inline .github/workflows/build-zip-file.yml]: Can we rename the file to `build-zip-and-run-smoke-tests.yml`, for instance?

And adjust the name l.1 accordingly

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [inline .github/workflows/build-zip-file.yml]:
```suggestion
# This action will run inside the package creation workflows or when it is triggered manually
```

File: .github/workflows/e2e-pull-request.yml
Comment (jbordonado) [inline .github/workflows/e2e-pull-request.yml]: I think, for now, we can remove the input and hardcode the default value in the below ENV variable.
We might never need it in the future, I don't see the target directory being ever different anytime soon, so my opinion is to keep it simple for a start

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [inline .github/workflows/release-code-freeze.yml]: We discussed a few weeks ago that it would make more sense to include this in the child workflow, i.e. `release-pr.yml`, since  we will catch not only regular releases but also hotfix releases (for which the release lead would manually trigger `release-pr` instead of `release-code-freeze`)

File: .github/workflows/e2e-pull-request.yml
Comment (jbordonado) [inline .github/workflows/e2e-pull-request.yml:34]: Do you need these env variables to access them within `setup.sh`? Are they not available from the inputs directly?

If not, can we then use the env variables everywhere to be consistent? (l.71 & l.75)

**Review by jbordonado — CHANGES_REQUESTED** [review]
I left a first round of comments, but I didn't test anything yet.
Let me know what you think, I'll start testing things out after the first changes (if you agree with some of my feedback hopefully 😄)

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [inline .github/workflows/build-zip-file.yml]: Can we use `! inputs.skip-smoke-tests` instead of `inputs.skip-smoke-tests != true`?
I remember I had some problems with booleans and if conditions back then, but I see you've been using both notations in the PR, so maybe it works nicely everywhere, and we can be consistent?

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: @jbordonado I understand it's always best to keep inputs to minimum. However, in this case, the caller workflow & build zip are 2 different workflows / jobs and hence to checkout a specific branch, that needs to be passed as an input.

I'm not sure whether it's possible to checkout a specific branch on one job/workflow & use it on another one without passing it as an input. For instance, the release branch is created from `develop` by the [release - code freeze workflow](https://github.com/Automattic/woocommerce-payments/actions/workflows/release-code-freeze.yml) and then calls the build zip workflow. If we don't pass the branch name to build zip, it would run on `develop`. 

Let me know your thoughts.

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: I added it as a use case if someone wishes to just create the release branch & PR, but not build a zip file, which is a rare case. Happy to remove if you think it's not required.

And yes, we want smoke tests to be run, so build zip is a per-requisite.

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: Ah yes it's not needed. I added the deletion step because somehow a build zip ended up inside the release folder, when I tried the build zip. It's not happening right now.

I'll remove it.

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [reply .github/workflows/build-zip-file.yml]: Yeah, I don't see when we wouldn't want a zip out of the release branch, to be honest. And if we do, we can just ignore the zip file run for this one occurrence. So I vote to remove this input

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [reply .github/workflows/build-zip-file.yml]: You're definitely right. I had something else in mind, but it looks like it's not possible actually.

File: .github/workflows/build-zip-file.yml
Comment (jbordonado) [reply .github/workflows/build-zip-file.yml]: Without forgetting to remove the "delete build zip" part of the step's name too

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: Removed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/06a924e2509b932eb181f15041d910b8d43d9a81

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: Removed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/06a924e2509b932eb181f15041d910b8d43d9a81

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: Removed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/b270fa50fc9a47d4041741b46c9374592d5e985b

File: .github/workflows/create-pre-release.yml
Comment (achyuthajoy) [reply .github/workflows/create-pre-release.yml]: Removed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/c1afe1a34626f1512c7d379928d473097e73f645

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: Updated in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/032c1da45c2e3e06c7d25fc05839e8fbeae5ed71 & https://github.com/Automattic/woocommerce-payments/pull/5586/commits/3bfe465a93a1f781e45928145acb65c735742135

File: .github/workflows/e2e-pull-request.yml
Comment (achyuthajoy) [reply .github/workflows/e2e-pull-request.yml]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/4dd1cdc983cae2b2a612cc644c5e72494da34c37

File: .github/workflows/e2e-pull-request.yml
Comment (achyuthajoy) [reply .github/workflows/e2e-pull-request.yml:34]: The setup script can't access the inputs directly hence we use env vars. I've updated the references

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml]: Moved build zip & smoke tests to `release-pr` workflow - https://github.com/Automattic/woocommerce-payments/pull/5586/commits/bdc6da56a0be26c268e42144e0229bebfef5360f

File: .github/workflows/build-zip-file.yml
Comment (achyuthajoy) [reply .github/workflows/build-zip-file.yml]: Updated in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/d15f4b18c70daa7a3431cb520699709a4e44a40e

**Comment by achyuthajoy** [comment]
@jbordonado I've address all the reviews. Good for another round of review.

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [inline .github/workflows/release-code-freeze.yml]: Is the goal to avoid skipping the Slack ping in case the `build-zip-and-run-smoke-tests` job added in `release-pr.yml` were to fail?

Can we look into the [continue-on-error](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error) job's prop for that? I assume we can achieve that goal by just specifying that property on `build-zip-and-run-smoke-tests` job, but it would need to be checked.

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [inline .github/workflows/release-code-freeze.yml:97]: Can we slightly update the text from the Slack ping? I believe we can make it more concise and clear, and we can also mention that we now already built the zip file + run the smoke tests.

Here is what I had in mind: [Slack Block Kit Builder preview](https://app.slack.com/block-kit-builder/T024FN1V2#%7B%22blocks%22:%5B%7B%22type%22:%22header%22,%22text%22:%7B%22type%22:%22plain_text%22,%22text%22:%22Code%20freeze%20notification%22,%22emoji%22:true%7D%7D,%7B%22type%22:%22section%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22The%20automated%20workflow%20just%20did%20the%20following%20actions:%22%7D%7D,%7B%22type%22:%22section%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22%E2%80%A2%20Create%20a%20release%20branch%20%60release/x.y.z%60%20%5Cn%20%E2%80%A2%20Raise%20a%20Pull%20Request%20using%20%60trunk%60%20as%20the%20base%20%5Cn%20%E2%80%A2%20Build%20a%20zip%20file%20and%20run%20the%20E2E%20tests%20against%20it%22%7D%7D,%7B%22type%22:%22section%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22If%20you%20absolutely%20need%20to%20include%20a%20PR%20that%20was%20not%20merged%20in%20%60develop%60%20yet,%20please%20consult%20the%20release%20lead.%22%7D%7D,%7B%22type%22:%22section%22,%22fields%22:%5B%7B%22type%22:%22mrkdwn%22,%22text%22:%22*Version:*%5Cn$%7B%7B%20env.RELEASE_VERSION%20%7D%7D%22%7D,%7B%22type%22:%22mrkdwn%22,%22text%22:%22*Planned%20date:*%5Cn%3Chttps://wcpay.wordpress.com/dev-resources/wc-payments-release-calendar/%7C$%7B%7B%20env.RELEASE_DATE%20%7D%7D%3E%22%7D%5D%7D%5D%7D)

Note: replace `release/x.y.z` by the corresponding env variables in the code, and remove `GITHUB_TAG_URL` env from the job, since it will not used anymore

Basically, keep it short with list items to enumerate what the workflow has done, and remove the link behind the `version` (the GH release doesn't exist yet when the ping is sent, so the link returns a 404).

Nice to have: 

- add a new section with 2 fields `Release zip file` and `Smoke tests run`, below the `Version` and `Planned date` section

OR (if not possible; not clear if we can retrieve a working link for the artifact before the workflow finishes)

- add a PR comment to the `release/x.y.z>trunk` PR with a link to the zip file and a link to the smoke tests run

I think it's best if we ship what you already built, so do you mind creating a new issue for this "nice to have" improvement?

**Review by jbordonado — CHANGES_REQUESTED** [review]
I've been able to test successfully:
- Release code freeze, with or without smoke tests (skip flag)
- Release PR, with or without smoke tests (skip flag)

They used the newly created release branch to build the zip file, as expected.
The E2E tests then used the uploaded artifact, as expected.

I could not test the `build-zip-file-and-run-smoke-tests` workflow via a manual trigger, since it doesn't exist yet, and I didn't want to bother with a hacky way (add `push` trigger first, and trigger through gh cli). But I don't see any issue in the code for the case where it is triggered manually (and where it should use the branch where we triggered it). That should work fine.

I just had one comment on the way to keep the Slack ping even though the E2E tests would fail. I believe we can use a built-in field to achieve this. I also suggested a few changes on the Slack ping content, which shouldn't require much work (since I shared the exact format through the Block Kit link)

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml]: @jbordonado `continue-on-error` would work in this case, but there is a problem with the approach.

* If the release PR is created and build zip or smoke tests failed, we can make use of `continue-on-error` to send out the slack notification.
* But if the release PR creation failed for any reason (e.gbranch already exists), using `continue-on-error` would still send out the slack notification even if the PR is not created. Hence I added an output to confirm that the PR was ready before sending the slack alert.

WDYT?

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml:97]: > add a new section with 2 fields Release zip file and Smoke tests run, below the Version and Planned date section

@jbordonado We can link to the GH action run, which would also help with downloading the artifact & see the status of smoke test. It won't be possible to get the artifact data until the workflow is finished. 

I tried fetching the artifact data within the workflow and I got `{"total_count":0,"artifacts":[]}`

But after the workflow is finished I got:

```
{
  "total_count": 1,
  "artifacts": [
    {
      "id": 592122679,
      "node_id": "MDg6QXJ0aWZhY3Q1OTIxMjI2Nzk=",
      "name": "woocommerce-payments",
      "size_in_bytes": 14341765,
      "url": "https://api.github.com/repos/Automattic/woocommerce-payments/actions/artifacts/592122679",
      "archive_download_url": "https://api.github.com/repos/Automattic/woocommerce-payments/actions/artifacts/592122679/zip",
      "expired": false,
      "created_at": "2023-03-10T04:26:27Z",
      "updated_at": "2023-03-10T04:26:28Z",
      "expires_at": "2023-03-24T04:25:39Z",
      "workflow_run": {
        "id": 4381298159,
        "repository_id": 173384525,
        "head_repository_id": 173384525,
        "head_branch": "devops/automate-smoke-tests",
        "head_sha": "bc3dd1955540d17d339d12d086146fa95194b42e"
      }
    }
  ]
}
```

Here is the GitHub documentation related to artifacts.
> During a workflow run, you can use the [download-artifact](https://github.com/actions/download-artifact) action to download artifacts that were previously uploaded in the same workflow run. After a workflow run has been completed, you can download or delete artifacts on GitHub or using the REST API.

Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml:97]: So the options we have are:

1. Add link to the workflow run in the slack notification which can be used to download the artifact & check smoke test status.
2. Move slack notification to a new workflow which can be triggered when required and runs after the current workflow is finished. Then we can fetch the artifact data and send it with the slack ping. I don't wish to complicate things more, so this is my least preferred option atm.

Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml:97]: I will create new issues to improve the slack notification. We can look at them later

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [reply .github/workflows/release-code-freeze.yml]: > But if the release PR creation failed for any reason (e.gbranch already exists), using continue-on-error would still send out the slack notification even if the PR is not created.

I thought we would add `continue-on-error: true` only on the `build-zip-and-run-smoke-tests` job inside `release-pr.yml`, meaning:
- the Slack job would run even when the `build-zip-and-run-smoke-tests` job inside `release-pr.yml` fails
- the Slack job would not run when the `prepare-release` job inside `release-pr.yml` fails, which is fine because we don't want to say on Slack that the release branch and PR are ready when it failed

But I realize now that this implies that if we trigger `release-pr.yml` manually and the smoke tests job fails, then we would not see this reflected on the workflow's run status, which may not be what we want (I'm not sure if that's important)

I'll let you decide if we keep it that way given these details

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [reply .github/workflows/release-code-freeze.yml:97]: Option 1 is fine. But as I said, all of these can happen later in a new set of issues, I just wanted to lay out my ideas when I started to review the PR.

If we can just adjust the wording like on the link I shared (and keep the PR comment, or new text sections for later), that would be already very good

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [inline .github/workflows/release-code-freeze.yml]: Using [gh pr comment](https://cli.github.com/manual/gh_pr_comment) directly could be easier and avoid introducing a 3rd party action:

`gh pr comment ${{ needs.create-release-pr.outputs.release-pr-id }} --body "Hi from GitHub CLI"`

Not sure if you're experimenting on that PR or if you plan also to ship this, I just wanted to mention again that I didn't expect this in this PR, so it's fine if you decide to revert this part.

Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml]: I was just experimenting stuff :) to see if it works.

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml]: Doesn't look good if we mark a failed job successful. So let's stick to my approach for now and we'll improve later

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml:97]: I've added the link to the workflow run on slack I've added the workflow run link to the slack notification.

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml]: `gh pr` command works fine. I've removed the action.

**Comment by achyuthajoy** [comment]
@jbordonado Good for another review. I've added the following:

* Updated the slack notification content
* Included link to workflow run on slack notification
* Included link to Pull request on slack notification
* Added step to create comment on PR with the workflow link

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [inline .github/workflows/release-code-freeze.yml]: We can revert to the original name `slack-notification`

File: .github/workflows/release-code-freeze.yml
Comment (jbordonado) [inline .github/workflows/release-code-freeze.yml]: Same, we can revert

File: .github/workflows/release-pr.yml
Comment (jbordonado) [inline .github/workflows/release-pr.yml]: This `echo` can be removed? I assume it was there during the testing

**Review by jbordonado — APPROVED** [review]
A few things to clean up after you moved the PR comment back to `release-pr`, but pre-approving now not to waste any more time.

I retested the code freeze workflow only, and it seems to work well.
Here is the Slack notification I received, for reference:

Screenshot 2023-03-10 at 11 41 38

Great work, I think this is a great improvement! 🥳

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/62038c903bca224966cac8a4f64d2afa6f546a7c

File: .github/workflows/release-code-freeze.yml
Comment (achyuthajoy) [reply .github/workflows/release-code-freeze.yml]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/62038c903bca224966cac8a4f64d2afa6f546a7c

File: .github/workflows/release-pr.yml
Comment (achyuthajoy) [reply .github/workflows/release-pr.yml]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/5586/commits/294f03bd554f2a0fdb88753af87032b0fb64cdc0

**Pull request merged**
